library/core/src/primitive_docs.rs

```diff
@@ -279,16 +279,44 @@ mod prim_never {}
 ///
 /// The `char` type represents a single character. More specifically, since
 /// 'character' isn't a well-defined concept in Unicode, `char` is a '[Unicode
-/// scalar value]', which is similar to, but not the same as, a '[Unicode code
-/// point]'.
-///
-/// [Unicode scalar value]: https://www.unicode.org/glossary/#unicode_scalar_value
-/// [Unicode code point]: https://www.unicode.org/glossary/#code_point
+/// scalar value]'.
 ///
 /// This documentation describes a number of methods and trait implementations on the
 /// `char` type. For technical reasons, there is additional, separate
 /// documentation in [the `std::char` module](char/index.html) as well.
 ///
+/// # Validity
+///
+/// A `char` is a '[Unicode scalar value]', which is any '[Unicode code point]'
+/// other than a [surrogate code point]. This has a fixed numerical definition:
+/// code points are in the range `'\0'` to `char::MAX` (`'\u{10FFFF}'`), inclusive.
+/// Surrogate code points, used by UTF-16, are in the range U+D800 to U+DFFF.
+///
+/// No `char` may be constructed, whether as a literal or at runtime, that is not a
+/// Unicode scalar value:
+///
+/// ```text
+/// let forbidden_chars = [
+///     // Each of these is a compiler error
```

Comment:
suggestion: split these into a couple separate doctests.  That way you'd be able to do things like

`````rust
/// ```rust,compile_fail
/// let _ = '\u{D800}';
/// ```
`````

rather than have them be `text`.

Reply:
Rustdoc does not like those literals; unless there's some trick I'm missing, they have to be in text mode. Using your example (including `compile_fail`) gives me an error a few blocks down:
```
> ./x.py doc library/std
...
   --> library/core/src/primitive_docs.rs:278:1
    |
278 | / /// A character type.
279 | | ///
280 | | /// The `char` type represents a single character. More specifically, since
281 | | /// 'character' isn't a well-defined concept in Unicode, `char` is a '[Unicode
...   |
379 | | /// assert_eq!(32, std::mem::size_of_val(&v[..]));
380 | | /// ```
    | |_______^
    |
    = note: `-D rustdoc::invalid-rust-codeblocks` implied by `-D warnings`
    = note: error from rustc: invalid unicode character escape

error: could not document `core`
```
and I tried several variants on that yesterday (including `ignore`). I can split the block up though.

Reply:
Oh, interesting.  I made sure to try it in play, where it seemed to work fine: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2a5d5de865ef87b2174cd33b553d2a46>

Maybe it's something you can `#[allow(rustdoc::invalid_rust_codeblocks)]`?  It'd be nice to get the "this intentionally doesn't compile" message next to it.

But it doesn't need to block the PR.  If it's a pain to get it to happen, the `text` is fine.

Reply:
I think that fixed it, thanks!